.grit/patterns/jest_to_vitest.md

```diff
@@ -0,0 +1,149 @@
+---
+title: Jest to Vitest
+---
+
+Convert Jest test to Vitest.
+
+tags: #migration, #js
+
+```grit
+engine marzano(0.1)
+language js
+
+pattern updated_imports() {
+  call_expression($function) where {
+    $function <: or {
+      js"test",
+      js"expect",
+      js"describe",
+      js"it",
+      js"beforeEach",
+      js"afterEach",
+      js"beforeAll",
+      js"afterAll"
+    } where {
+      $source = `'vitest'`,
+      $function <: ensure_import_from($source),
+      $vi = `vi`,
+      $vi <: ensure_import_from($source)
+    }
+  }
+}
+
+or {
+  // Imports
+  updated_imports(),
+  `import $_ from 'jest'` => .,
+
+  `jest.requireActual($module)` => `await vi.importActual($module)`,
+
+  `JEST_WORKER_ID` => `VITEST_POOL_ID`,
+
+  // module mocks
+  `jest.mock($module, $mockImplementation)` where { 
+    $mockImplementation <: arrow_function($body) where { 
+      $body <: literal_value() => `({ 
+        default: $body 
+      })`
```

Comment:
The body could be a lot more than just a literal value for it to need to be modified in this way. This seems too narrow.

Reply:
Hi @morgante 
From the document: `the factory argument's return value is the default export`
if the mock implementation already have `default` in return object or it's for mocking export only (not default) then wrapping them in `default` is not correct

Reply:
From the document: `the factory argument's return value is the default export`
if the mock implementation already have `default` in return object or it's for mocking export only (not default) then wrapping them in `default` is not correct

Reply:
Don't just quote the document back to me.

I'm saying this pattern is not nearly general enough. We need to handle more than just literal values. If you added more test cases, you would see that.

For example, you have not covered this case that should clearly be handled:

```
jest.mock("something", () => {
  someCounter: () => { return 1; }
});
```

Reply:
Thanks @morgante for the review. For collecting `jest.mock` use cases and marking sure which kind of migrations are correct with them, we need some times to setup a lighter repo with multiple/runnable `jest.mock` usages and do the migrations there to make sure it works perfectly. We would update this thread in next commits

Reply:
@morgante
I setup a repo with some `jest.mock` here (https://github.com/phunguyenmurcul/node-typescript-jest-playground/tree/main/__tests__) and jest test works fine.

I made a PR for jest to vitest migration: https://github.com/phunguyenmurcul/node-typescript-jest-playground/pull/2

`vitest` works well with correct mocking setups and I don't need to make any changes in mock implementation.

One of those mocks is similar to your sample for `should clearly be handled` case in your last comment.
What do you think about this?

Reply:
Your example is relying on `export const`. Please try to see how it works when the module is *only* relying on an object default export.

```
export default {
  methodFoo: (): string => {
    return 'foo';
  },
  methodBar = (): string => {
  return 'bar';
  }
}
```

Reply:
you can see in this [test file](https://github.com/phunguyenmurcul/node-typescript-jest-playground/blob/main/__tests__/mock-default.test.ts), I mock the `default` export in [this module](https://github.com/phunguyenmurcul/node-typescript-jest-playground/blob/main/src/mock.ts#L11C16-L11C23). Is it the setup you want to see?

P/s: without [this `default`](https://github.com/phunguyenmurcul/node-typescript-jest-playground/blob/main/__tests__/mock-default.test.ts#L4C3-L4C11), `jest` doesn't work

Reply:
Ok I did some of my own validation [here](https://github.com/phunguyenmurcul/node-typescript-jest-playground/compare/main...morgante:node-typescript-jest-playground:jest-mock) and it seems to work.